Zend/zend_gc.c

```diff
@@ -710,28 +779,33 @@ static void gc_scan_black(zend_refcounted *ref)
 	ref = Z_COUNTED_P(zv);
 	GC_ADDREF(ref);
 	if (!GC_REF_CHECK_COLOR(ref, GC_BLACK)) {
+		GC_REF_SET_BLACK(ref);
+		goto tail_call;
+	}
+
+next:
+	ref = GC_STACK_POP();
+	if (ref) {
 		goto tail_call;
 	}
 }
 
-static void gc_mark_grey(zend_refcounted *ref)
+static void gc_mark_grey(zend_refcounted *ref, gc_stack *stack)
 {
-    HashTable *ht;
+    HashTable *ht = NULL;
 	Bucket *p, *end;
 	zval *zv;
+	GC_STACK_DCL(stack);
 
-tail_call:
-	if (!GC_REF_CHECK_COLOR(ref, GC_GREY)) {
-		ht = NULL;
+	do {
```

Comment:
I prefer to avoid "goto", if possible. I kept them in other places to minimize the patch.
Probably I'll re-factor the code in PHP-7.4 and master after merging this.

Reply:
From what I notices it's just to prevent white-space changes. I did tricks like that also.